File: Sources/SwiftCogCore/Kernels/MotorKernel.swift
import Foundation
import Distributed
import DistributedCluster

public distributed actor MotorKernel: Kernel {
    let kernelId: KernelID
    unowned let system: KernelSystem
    private let customHandler: ((KernelMessage, MotorKernel) async throws -> Void)?

    public init(actorSystem: DefaultDistributedActorSystem, system: KernelSystem, customHandler: ((KernelMessage, MotorKernel) async throws -> Void)? = nil) {
        self.actorSystem = actorSystem
        self.kernelId = KernelID()
        self.system = system
        self.customHandler = customHandler
    }

    public distributed func getKernelId() -> KernelID {
        return self.kernelId
    }

    public distributed func receive(message: KernelMessage) {
        Task {
            if let customHandler = customHandler {
                // Use the custom handler if provided
                try await customHandler(message, self)
            } else {
                // Default behavior: pass the message on
                try await system.emit(message: message, from: self)
            }
        }
    }
} 

File: Sources/server/motor_kernel.py
"""
Motor kernel implementation for the SwiftCog Python server.
"""
from typing import Callable, Optional
import ray
from swiftcog_types import KernelID, KernelMessage


@ray.remote
class MotorKernel:
    """Motor kernel implementation matching the Swift version."""
    
    def __init__(self, custom_handler: Optional[Callable] = None):
        self.custom_handler = custom_handler
        self.kernel_id = KernelID.MOTOR
    
    def get_kernel_id(self) -> KernelID:
        return self.kernel_id
    
    async def receive(self, message: KernelMessage) -> None:
        """Receive and process a message."""
        if self.custom_handler:
            await self.custom_handler(message, self)
        else:
            await self.default_handler(message)
    
    async def default_handler(self, message: KernelMessage) -> None:
        """Default handler that processes motor commands and forwards directly to Expression."""
        print(f"MotorKernel: Processing {message.payload}")
        
        # Hardcoded connection: Motor -> Expression
        try:
            expression_kernel = ray.get_actor("ExpressionKernel")
            await expression_kernel.receive.remote(message)
            print("MotorKernel -> ExpressionKernel")
        except ValueError:
            print("Error: ExpressionKernel not found") 